server/server.go

```diff
@@ -0,0 +1,393 @@
+			writer.CloseWithError(err)
+			break
+		}
+		switch payload := req.GetPayload().(type) {
+		case *pb.AddFileRequest_Chunk:
+			_, writeErr := writer.Write(payload.Chunk)
+			if writeErr != nil {
+				return writeErr
+			}
+		default:
+			return fmt.Errorf("expected Chunk for AddFileRequest.Payload but got %T", payload)
+		}
+	}
+
+	addFileResult := <-addFileChannel
+	if addFileResult.Err != nil {
+		return addFileResult.Err
+	}
+
+	respNode, err := nodeToPbNode(addFileResult.Node)
+	if err != nil {
+		return err
+	}
+
+	return srv.SendAndClose(&pb.AddFileResponse{Node: respNode})
+}
+
+func (s *ipfsLiteServer) GetFile(req *pb.GetFileRequest, srv pb.IpfsLite_GetFileServer) error {
+	cid, err := cid.Decode(req.GetCid())
+	if err != nil {
+		return err
+	}
+
+	reader, err := s.peer.GetFile(srv.Context(), cid)
+	if err != nil {
+		return err
+	}
+
+	buffer := make([]byte, getFileChunkSize)
+	for {
+		size, err := reader.Read(buffer)
+		if err != nil && err != io.EOF {
+			return err
+		}
+		srv.Send(&pb.GetFileResponse{Chunk: buffer[:size]})
+		if err == io.EOF {
+			return nil
+		}
+	}
+}
+
+func (s *ipfsLiteServer) HasBlock(ctx context.Context, req *pb.HasBlockRequest) (*pb.HasBlockResponse, error) {
+	cid, err := cid.Decode(req.GetCid())
+	if err != nil {
+		return nil, err
+	}
+	hasBlock, err := s.peer.HasBlock(cid)
+	if err != nil {
+		return nil, err
+	}
+	return &pb.HasBlockResponse{HasBlock: hasBlock}, nil
+}
+
+func (s *ipfsLiteServer) AddNode(ctx context.Context, req *pb.AddNodeRequest) (*pb.AddNodeResponse, error) {
+	cid, err := cid.Decode(req.Block.GetCid())
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: there is also just blocks.NewBlock()
+	block, err := blocks.NewBlockWithCid(req.Block.GetRawData(), cid)
+	if err != nil {
+		return nil, err
+	}
+
+	node, err := format.Decode(block)
+	if err != nil {
+		return nil, err
+	}
+
+	err = s.peer.Add(ctx, node)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.AddNodeResponse{}, nil
+}
+
+func (s *ipfsLiteServer) AddNodes(ctx context.Context, req *pb.AddNodesRequest) (*pb.AddNodesResponse, error) {
+	nodes := make([]format.Node, len(req.GetBlocks()))
+
+	for i, pbBlock := range req.GetBlocks() {
+		cid, err := cid.Decode(pbBlock.GetCid())
+		if err != nil {
+			return nil, err
+		}
+
+		// TODO: there is also just blocks.NewBlock()
+		block, err := blocks.NewBlockWithCid(pbBlock.GetRawData(), cid)
+		if err != nil {
+			return nil, err
+		}
+
+		node, err := format.Decode(block)
+		if err != nil {
+			return nil, err
+		}
+
+		nodes[i] = node
+	}
+
+	err := s.peer.AddMany(ctx, nodes)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.AddNodesResponse{}, nil
+}
+
+func (s *ipfsLiteServer) GetNode(ctx context.Context, req *pb.GetNodeRequest) (*pb.GetNodeResponse, error) {
+	cid, err := cid.Decode(req.GetCid())
+	if err != nil {
+		return nil, err
+	}
+	// TODO: use session() NodeGetter or Peer NodeGetter methods directly?
+	node, err := s.peer.Get(ctx, cid)
```

Comment:
I'm calling the peer's get node method directly, but there is also a `Session()` method that returns a "session-based NodeGetter". Not sure what the difference is or which I should use.

Reply:
Cool, good enough for now. I've never used the sessions but they could be handy at some point. The idea is to reduce bandwidth usage if you need to get a bunch of dags (I think): 

From the `go-blockservice`
```
// NewSession creates a new session that allows for
// controlled exchange of wantlists to decrease the bandwidth overhead.
// If the current exchange is a SessionExchange, a new exchange
// session will be created. Otherwise, the current exchange will be used
// directly.
```

Reply:
Sounds good.